dimod/include/dimod/adjvectordqm.h

```diff
@@ -0,0 +1,356 @@
+        variable_type num_total_cases = bqm_.num_variables();
+
+        for (variable_type v = 0; v < num_variables; v++) {
+            case_starts[v] = case_starts_[v];
+        }
+
+        size_type qi = 0;
+        for (variable_type ci = 0; ci < num_total_cases; ci++) {
+            linear_biases[ci] = bqm_.get_linear(ci);
+            auto span = bqm_.neighborhood(ci);
+            while ((span.first != span.second) && ((span.first)->first < ci)) {
+                irow[qi] = ci;
+                icol[qi] = (span.first)->first;
+                quadratic_biases[qi] = (span.first)->second;
+                span.first++;
+                qi++;
+            }
+        }
+    }
+
+    bool self_loop_present() {
+        variable_type num_variables = this->num_variables();
+        for (variable_type v = 0; v < num_variables; v++) {
+            for (variable_type ci = case_starts_[v],
+                               ci_end = case_starts_[v + 1];
+                 ci < ci_end; ci++) {
+                auto span = bqm_.neighborhood(ci, case_starts_[v]);
+                if ((span.first != span.second) &&
+                    ((span.first)->first < case_starts_[v + 1])) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    bool connection_present(variable_type u, variable_type v) {
+        bool connected = true;
+        auto it = std::lower_bound(adj_[u].begin(), adj_[u].end(), v);
+        if (it == adj_[u].end() || *it != v) {
+            connected = false;
+        }
+        return connected;
+    }
+
+    void connect_variables(variable_type u, variable_type v) {
+        auto low = std::lower_bound(adj_[u].begin(), adj_[u].end(), v);
+        if (low == adj_[u].end() || *low != v) {
+            adj_[u].insert(low, v);
+            adj_[v].insert(std::lower_bound(adj_[v].begin(), adj_[v].end(), u),
+                           u);
+        }
+    }
+
+    size_type num_variables() { return adj_.size(); }
+
+    size_type num_variable_interactions() {
+        size_type num = 0;
+        variable_type num_variables = this->num_variables();
+        for (variable_type v = 0; v < num_variables; v++) {
+            num += adj_[v].size();
+        }
+        return (num / 2);
+    }
+
+    size_type num_cases(variable_type v) {
+        assert(v >= 0 && v < this->num_variables());
+        return (case_starts_[v + 1] - case_starts_[v]);
+    }
+
+    size_type num_case_interactions() { return bqm_.num_interactions(); }
+
+    variable_type add_variable(variable_type num_cases) {
+        assert(num_cases > 0);
+        variable_type v = adj_.size();
+        adj_.resize(v + 1);
+        for (variable_type n = 0; n < num_cases; n++) {
+            bqm_.add_variable();
+        }
+        case_starts_.push_back(bqm_.num_variables());
+        return v;
+    }
+
+    bias_type get_linear_case(variable_type v, variable_type case_v) {
```

Comment:
In the bqm I am switching to `bias_type& linear(v)` and I think `bias_type& linear_case(u, v)` would also be preferable to get/set.

Note that for setting to work with cython, you need some trickery. E.g.
```
        cdef bias_type *b = &(self.cppbqm.linear(v))
        b[0] = bias
```
In python there isn't really a notion of a reference/proxy object in the same way. So we'll want to keep get/set.